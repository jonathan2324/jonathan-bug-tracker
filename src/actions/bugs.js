import database from '../firebase/firebase'

//ADD_BUG
export const addBug = (bug) => ({
    type: 'ADD_BUG',
    bug
    
})

//START addBUg
export const startAddBug = (bugData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const { 
            name = '',
            description = '',
            notes = '',
            priority = 'Low',
            completed = false,
            createdAt = 0,
            contributions = [{id: '0', name: "initial", contribution: "initial"}]
          } = bugData

          const bug = { name, description, priority, completed, notes, createdAt, contributions }

          return database.ref(`users/${uid}/bugs`).push(bug).then((ref) => {
              dispatch(addBug({
                  id: ref.key,
                  ...bug
              }))
          })
    }
}

//REMOVE_BUG

export const removeBug = ({ id }) => ({
    type: 'REMOVE_BUG',
    id
})

//Start REMOVE_BUG
export const startRemoveBug = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/bugs/${id}`).remove().then(() => {
            dispatch(removeBug({ id }))
        })
    }
}

//Edit Bug
export const editBug = (id, updates) => ({
    type: 'EDIT_BUG',
    id,
    updates
})

//Start edit bug
export const startEditBug = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/bugs/${id}`).update(updates).then(() => {
            dispatch(editBug(id, updates))
        })
    }
}

//TOGGLE COMPLETED
export const toggleCompletedBug = (id, completed) => ({
    type: 'TOGGLE_COMPLETED_BUG',
    id,
    completed
})

//start Toggle Completed

export const startToggleCompletedBug = (id, completed) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/bugs/${id}`).update({ completed: !completed }).then(() => {
            dispatch(toggleCompletedBug(id, completed))
        })
    }
}

//ADD_CONTRIBUTION
export const addContribution = (id, contribution) => ({
    type: 'ADD_CONTRIBUTION',
    id,
    contribution

})

//Start add CONTRIBUTION
export const startAddContribution = (id, bugContributionData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const { 
            name = '',
            contribution = ''
          } = bugContributionData

          const contributionData = { name, contribution }
          return database.ref(`users/${uid}/bugs/${id}/contributions`).push(contributionData).then((ref) => {
              dispatch(addContribution(id, {
                  id: ref.key,
                  name: contributionData.name,
                  contribution: contributionData.contribution
              }))
          })
    }
}


//EDIT_CONTRIBUTION
export const editContribution = (bugID, contributionID, updatedContribution) => ({
    type: 'EDIT_CONTRIBUTION',
    bugID,
    contributionID,
    updatedContribution
})

//Start edit contribution
export const startEditContribution = (bugID, contributionID, updatedContribution) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/bugs/${bugID}/contributions/${contributionID}`).update(updatedContribution).then(() => {
            dispatch(editContribution(bugID, contributionID, updatedContribution))
        })
    }
}
//REMOVE_CONTRIBUTION
export const removeContribution = (bugID, contributionID) => ({
    type: 'REMOVE_CONTRIBUTION',
    bugID,
    contributionID
})

//Start remove contribution
export const startRemoveContribution = (bugID, contributionID) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/bugs/${bugID}/contributions/${contributionID}`).remove().then(() => {
            dispatch(removeContribution(bugID, contributionID))
        })
    }
}

//SET_BUGS
export const setBugs = (bugs) => ({
    type: 'SET_BUGS',
    bugs
})

//startSetBugs

export const startSetBugs = (bugs) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        //return here makes sure the promise gets returned
        return database.ref(`users/${uid}/bugs`).once('value').then((snapshot) => {
            const bugs = []
            snapshot.forEach((childSnapshot) => {
                bugs.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setBugs(bugs))
        })

        
    }
}


//set contributions
export const setContributions = (id, contributions = []) => (    
    {
    type: 'SET_CONTRIBUTIONS',
    id,
    contributions
    
})

//start set contributions
export const startSetContributions = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
         return database.ref(`users/${uid}/bugs/${id}/contributions`).once('value').then((snapshot) => {
            const contributions = []
            snapshot.forEach((childSnapshot) => {
                contributions.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()

                })
            })
    
            dispatch(setContributions(id, contributions))
        })
    }
}