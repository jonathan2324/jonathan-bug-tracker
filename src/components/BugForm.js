import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'



const now = moment()
console.log(now.format('MMM Do YYYY'))

class BugForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            name: props.bug ? props.bug.name : '',
            priority: props.bug ? props.bug.priority : 'Low',
            description: props.bug ? props.bug.description : '',
            notes: props.bug ? props.bug.notes : '',
            createdAt: props.bug ? moment(props.bug.createdAt) : moment(),
            calendarFocused: false,
            error: ''
    
        }
    }
    

    onNameChange = (e) => {
        const name = e.target.value 
        this.setState(() => ({ name }))
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNotesChange = (e) => {
        const notes = e.target.value
        this.setState(() => ({ notes }))       
    }

    onPriorityChange = (e) => {
        const priority = e.target.value
        this.setState(() => ({ priority }))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }))

        }

    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.name || !this.state.description) {
            this.setState(() => ({ error: 'Please provide an employee name and a description'}))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                notes: this.state.notes,
                name: this.state.name,
                createdAt: this.state.createdAt.valueOf(),
                priority: this.state.priority
            })
        }
    }

    render () {
        return (
            <div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmit}>
                    <input type='text' placeholder='Employee name' autoFocus value={this.state.name} onChange={this.onNameChange}/>
                    <select value={this.state.priority} onChange={this.onPriorityChange}>
                        <option value="High">High</option>
                        <option value="Med">Med</option>
                        <option value="Low">Low</option>
                    </select>

                    <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    />

                    <textarea type='text' maxLength="50" placeholder='Enter error description (max length 50 characters)' value={this.state.description} onChange={this.onDescriptionChange}/>
                    <textarea type='text' placeholder='Enter important notes about error' value={this.state.notes} onChange={this.onNotesChange}/>
                    <button>Add Bug</button>

                </form>
            </div>
        )
    }
}

export default BugForm