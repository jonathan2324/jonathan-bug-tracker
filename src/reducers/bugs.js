

//bugReducer
const bugReducerDefaultState = []

const bugReducer = (state = bugReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BUG':
            return [
                ...state,
                action.bug
            ]
        case 'REMOVE_BUG':
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        ...action.updates
                    }
                } else {
                    return bug
                }

            })
        case 'TOGGLE_COMPLETED_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        completed: !action.completed
                    }
                } else {
                    return bug
                }
            })
        case 'ADD_CONTRIBUTION':
            return state.map((bug) => {
                if(bug.id === action.id){
                    return {
                        ...bug,
                        contributions: [
                            
                            action.contribution,
                            ...bug.contributions,
                            
                        ]
                    }
                } else return bug
            })
        default:
            return state;
    }
};

export default bugReducer