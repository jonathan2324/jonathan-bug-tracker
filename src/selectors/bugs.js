import moment from 'moment'

//Get visible bugs
const getVisibleBugs = (bugs, { text, sortBy, startDate, endDate }) => {
    return bugs.filter((bug) => {
        const createdAtMoment = moment(bug.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = bug.name.toLowerCase().includes(text.toLowerCase())
        
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else {
            return null
        }
    }).sort((a,b) => {
       if(sortBy === 'completed') {
           return a.completed ? 1 : -1
       } else {
           return 0
       }
    }).sort((a, b) => {
        if(sortBy === 'priority') {
            if(a.priority === 'High') {
                return b.priority === 'Med' || b.priority === 'Low' ? -1 : 1
            }
            else if (a.priority === 'Med') {
                return b.priority === 'High' ? 1 : -1
            } else {
                return 0
            }
        }
            
    })
}

export default getVisibleBugs