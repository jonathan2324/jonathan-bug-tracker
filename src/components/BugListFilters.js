import React from 'react'
import {connect} from 'react-redux'
import { setTextFilter, sortByDate, sortByPriority, setStartDate, setEndDate, sortByCompleted, filterDefault, filterByCompleted, filterByUncompleted } from '../actions/filters'
import { DateRangePicker } from 'react-dates'
import getVisibleBugs from '../selectors/bugs'



export class BugListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => { //destructure object that is passed into this function by the date range picker
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)        //these are the startDate and endDate that was required by airbnb
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused: calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
        }
    
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        } else if (e.target.value === 'priority') {
            this.props.sortByPriority()
        } 
    }

    onFilterChange = (e) => {
        if(e.target.value === 'uncompleted') {
            this.props.filterByUncompleted()
        } else if (e.target.value === 'completed') {
            this.props.filterByCompleted()
        } else if (e.target.value === 'all') {
            this.props.filterDefault()
        }
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='priority'>Priority</option>
                    
                </select>
                <select value={this.props.filters.filterBy} onChange={this.onFilterChange}>
                    <option value='all'>View all</option>
                    <option value='uncompleted'>View open</option>
                    <option value='completed'>View closed</option>
                </select>
                <DateRangePicker 
                startDate={this.props.filters.startDate} 
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
                />
            </div>
        )
    }

}

const mapStateToProps = (state) => {
    const visibleBugs = getVisibleBugs(state.bugs, state.filters)
    return {
        filters: state.filters,
        completedBugs: visibleBugs.filter((bug) => { return bug.completed === 'true'}),
        uncompletedBugs: visibleBugs.filter((bug) => { return bug.completed === 'false'})
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByCompleted: () => dispatch(sortByCompleted()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    filterDefault: () => dispatch(filterDefault()),
    filterByCompleted: () => dispatch(filterByCompleted()),
    filterByUncompleted: () => dispatch(filterByUncompleted())
})

export default connect(mapStateToProps, mapDispatchToProps)(BugListFilters)