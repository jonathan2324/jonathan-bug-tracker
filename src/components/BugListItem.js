import React from 'react'
import { Link } from 'react-router-dom'
import {connect} from 'react-redux'
import { startToggleCompletedBug } from '../actions/bugs'
import moment from 'moment'

export const BugListItem = ({ dispatch, id, name, description, priority, createdAt, completed}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>
                {name}
            </h3>
        </Link>
            <p>{description}</p>
            
            <p>{priority}</p>

            <p>
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        

        <button onClick={(e) => dispatch(startToggleCompletedBug(id, completed))}>{completed ? 'Closed': 'Open'}</button>
    </div>


)


export default connect()(BugListItem)