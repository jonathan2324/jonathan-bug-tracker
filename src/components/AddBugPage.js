import React from 'react';
import BugForm from './BugForm'
import { connect } from 'react-redux'
import { addBug } from '../actions/bugs'

export class AddBugPage extends React.Component {
  onSubmit = (bug) => {
    this.props.addBug(bug)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <h1>Add bug</h1>
        <BugForm onSubmit={this.onSubmit}/>
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    addBug:(bug) => dispatch(addBug(bug))
  }
}

export default connect(undefined, mapDispatchToProps)(AddBugPage);