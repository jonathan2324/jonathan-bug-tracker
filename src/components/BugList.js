import React from 'react'
import { connect } from 'react-redux'
import BugListItem from './BugListItem'
import selectBugs from '../selectors/bugs'

export const BugList = (props) => (
    <div>
        {props.uncompletedBugs.length > 0 && <h3>Uncompleted Bugs [{props.uncompletedBugs.length}]:</h3>}
        {
            props.bugs.length === 0 ? (

                <p>No bugs</p>

            ) : (
            
                props.uncompletedBugs.map((bug) => {
                    return <BugListItem key={bug.id} {...bug}/>
                 })

            )
        }
        {props.completedBugs.length > 0 && <h3>Completed bugs [{props.completedBugs.length}]:</h3>}
        {
            props.bugs.length === 0 ? (

                <p></p>

            ) : (
            
                props.completedBugs.map((bug) => {
                    return <BugListItem key={bug.id} {...bug}/>
                 })

            )
        }



    </div>
)

const mapStateToProps = (state) => {
    const visibleBugs = selectBugs(state.bugs, state.filters)

    return {
        bugs: visibleBugs,
        completedBugs: visibleBugs.filter((bug) => {return bug.completed === true}),
        uncompletedBugs: visibleBugs.filter((bug) => {return bug.completed === false})

    }
}

const ConnectedBugList = connect(mapStateToProps)(BugList)

export default ConnectedBugList