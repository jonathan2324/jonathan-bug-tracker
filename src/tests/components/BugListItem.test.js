import React from 'react'
import toJSON from 'enzyme-to-json'
import { shallow } from 'enzyme'
import bugs from '../fixtures/bugs'
import { BugListItem } from '../../components/BugListItem'



test('Should render bug list items', () => {
    const wrapper = shallow(<BugListItem {...bugs[0]}/>) //spread out an object so all its properties become props
    expect(toJSON(wrapper)).toMatchSnapshot()
})