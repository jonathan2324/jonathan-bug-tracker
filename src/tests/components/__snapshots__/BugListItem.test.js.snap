// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render bug list items 1`] = `
<div>
  <[object Object]
    to="/edit/1"
  >
    <h3>
      Lebron
    </h3>
    <p>
      Error in X
    </p>
    <p>
      Med
    </p>
    <p>
      December 31st, 1969
    </p>
  </[object Object]>
  <button
    onClick={[Function]}
  >
    Closed
  </button>
</div>
`;
